### PR TITLE
Lvruby 14: Added categories.

### PR DESCRIPTION
**Pull request by nataliakozoriz**
## JIRA ticket

https://ssu-jira.softserveinc.com/browse/LVRUBY-14

- [x] @okohuch 
- [x] @kmandziy

_Please list your code reviewers here._

## Summary of issue

Add categories.

## Summary of change

Added category model.
Added category controller.

## Review checklist

- [x] Rubocop doesn't raise any warnings and offenses
- [x] New features are covered with new specs
- [x] All specs pass

### CODE REVIEW

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **90.0%** covered at **1.4** hits/line

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -13,3 +13,26 @@
...
+button.btn {
+	margin: 10px;
+  &.btn-primary a {
+    color: white;
+    font-size: 24px;
+    &:hover {
+      text-decoration: none;
+    }
+  }
+}
```
Comment (kmandziy) [inline app/assets/stylesheets/application.scss]: Add end line

File: spec/factories/users.rb [hunk 2]
```diff
@@ -2,5 +2,7 @@
 
 FactoryBot.define do
   factory :user do
+    email { 'user@example.com' }
+    password { 'password' }
```
[the remark this answers is not in the document]
Comment (nataliakozoriz) [reply spec/factories/users.rb]: Done.

File: spec/controllers/categories_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'factory_bot'
+
+def login_user
+  before do
+    user = FactoryBot.create(:user)
```
[the remark this answers is not in the document]
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Done.

File: app/assets/stylesheets/application.scss (hunk 1 above)
Comment (nataliakozoriz) [reply app/assets/stylesheets/application.scss]: Done.

**Comment by okohuch** [comment]
fix circleci errors

File: app/helpers/categories_helper.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+module CategoriesHelper
+  def category_params
+    params.require(:category).permit(:name, :desc)
```
Comment (okohuch) [inline app/helpers/categories_helper.rb]: this must be located in the controller.
helpers are for the views

File: app/controllers/categories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    @category.destroy
+    redirect_to categories_path, notice: 'Delete success'
+  end
+
+  def edit
+    @category = Category.find(params[:id])
+  end
+
+  def update
+    @category = Category.find(params[:id])
```
Comment (okohuch) [inline app/controllers/categories_controller.rb]: use filters to assign `@category` variable

File: app/controllers/categories_controller.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    @category.update(category_params)
```
Comment (okohuch) [inline app/controllers/categories_controller.rb]: what is going to happen in case category was not updated? we should handle this case as well

File: app/controllers/categories_controller.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    if @category.save
+      redirect_to categories_path, flash: { success: 'Category was successfully created.' }
+    else
+      render 'new'
+    end
+  end
+
+  def destroy
+    @category = Category.find(params[:id])
+    @category.destroy
```
Comment (okohuch) [inline app/controllers/categories_controller.rb]: what if the category was not destroyed? we should handle it as well

File: db/migrate/20190413204542_create_categories.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateCategories < ActiveRecord::Migration[5.2]
+  def change
+    create_table :categories do |t|
+      t.string :name
+      t.text :desc
```
Comment (okohuch) [inline db/migrate/20190413204542_create_categories.rb]: use the full word `description`

File: config/database.yml.example [hunk 9]
```diff
@@ -17,3 +17,4 @@ test:
 production:
   <<: *default
   database: production_ai_colaboration_platform
+
```
Comment (okohuch) [inline config/database.yml.example]: update sublime settings and add new line at the end of file

File: spec/controllers/categories_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    end
+
+    it 'routes /categories/1 to categories#destroy' do
+      expect(delete: '/categories/1').to route_to(controller: 'categories',
+                                                  action: 'destroy', id: '1')
+    end
+  end
+end
+
+RSpec.describe CategoriesController, type: :controller do
```
Comment (okohuch) [inline spec/controllers/categories_controller_spec.rb]: you already described the controller in the 13th line

File: spec/controllers/categories_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,134 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'factory_bot'
+
+def login_user
```
Comment (okohuch) [inline spec/controllers/categories_controller_spec.rb]: do not create additional methods
use `before`

File: spec/controllers/categories_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    let(:category) { create :category }
+    it 'should success and render to edit page' do
+      get :show, params: { id: category.id }
+      expect(response).to have_http_status(200)
+      expect(response).to render_template :show
+    end
+  end
+end
+
+RSpec.describe CategoriesController, type: :controller do
```
Comment (okohuch) [inline spec/controllers/categories_controller_spec.rb]: you already described the controller in the 13th line

File: spec/controllers/categories_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        name: 'An awesome category',
+        desc: 'An awesome category for students'
+      }
+      expect { post(:create, params: { category: params }) }.to change(Category, :count).by(1)
+      expect(flash[:success]).to eq 'Category was successfully created.'
+    end
+  end
+end
+
+RSpec.describe CategoriesController, type: :controller do
```
Comment (okohuch) [inline spec/controllers/categories_controller_spec.rb]: you already described the controller in the 13th line

File: spec/controllers/categories_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+      put :update, params: { id: category.id, category: params }
+      category.reload
+      params.keys.each do |key|
+        expect(category.attributes[key.to_s]).to eq params[key]
+      end
+    end
+  end
+end
+
+RSpec.describe CategoriesController, type: :controller do
```
Comment (okohuch) [inline spec/controllers/categories_controller_spec.rb]: you already described the controller in the 13th line

File: spec/controllers/categories_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    let!(:category) { create :category }
+
+    it 'should delete category' do
+      expect { delete :destroy, params: { id: category.id } }.to change(Category, :count).by(-1)
+      expect(flash[:notice]).to eq 'Delete success'
+    end
+  end
+end
+
+RSpec.describe CategoriesController, type: :controller do
```
Comment (okohuch) [inline spec/controllers/categories_controller_spec.rb]: you already described the controller in the 13th line

File: spec/factories/users.rb [hunk 16]
```diff
@@ -2,5 +2,8 @@
 
 FactoryBot.define do
   factory :user do
+    email { 'user@example.com' }
```
Comment (okohuch) [inline spec/factories/users.rb]: use sequence for the email

File: spec/factories/users.rb [hunk 17]
```diff
@@ -2,5 +2,8 @@
...
+    password { 'password' }
+    confirmed_at { Time.new }
```
Comment (okohuch) [inline spec/factories/users.rb]: read about `Time.current` and why we should use it.
change it to `Time.current`

**Comment by vikabelous** [comment]
**Code Coverage:**
:warning: **88.1%** covered at **1.37** hits/line

File: app/controllers/categories_controller.rb (hunk 6 above)
Comment (nataliakozoriz) [reply app/controllers/categories_controller.rb]: Added else statement if not successful.

File: app/controllers/categories_controller.rb (hunk 7 above)
Comment (nataliakozoriz) [reply app/controllers/categories_controller.rb]: Added else statement if not successful.

File: config/database.yml.example (hunk 9 above)
Comment (nataliakozoriz) [reply config/database.yml.example]: Done.

File: spec/controllers/categories_controller_spec.rb (hunk 12 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Done one description of controller.

File: spec/controllers/categories_controller_spec.rb (hunk 13 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Done one description of controller.

File: spec/controllers/categories_controller_spec.rb (hunk 14 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Done one description of controller.

File: spec/controllers/categories_controller_spec.rb (hunk 15 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Done one description of controller.

File: spec/controllers/categories_controller_spec.rb (hunk 11 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Used before for signing in user.

File: spec/controllers/categories_controller_spec.rb (hunk 10 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Done one description of controller.

**Comment by vikabelous** [comment]
**Code Coverage:**
:warning: **88.1%** covered at **1.4** hits/line

File: app/assets/stylesheets/application.scss [hunk 18]
```diff
@@ -13,3 +13,26 @@
...
+
+html input[type="submit"] {
+  border: none;
+  background: none;
+  font-size: 24px;
+  color: white;
+}
+
+button.btn {
+	margin: 10px;
```
Comment (vikabelous) [inline app/assets/stylesheets/application.scss]: Fix indentations

File: app/models/category.rb [hunk 19]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class Category < ApplicationRecord
+  has_many :news
+  validates :name, presence: true, length: { minimum: 2 }
```
Comment (vikabelous) [inline app/models/category.rb]: Can we have two categories with the same name?

File: app/views/categories/edit.html.haml [hunk 20]
```diff
@@ -0,0 +1,15 @@
+%h1 Edit selected category
```
Comment (vikabelous) [inline app/views/categories/edit.html.haml]: You have problems with formatting
Add haml linter to your projects
https://github.com/brigade/haml-lint

File: app/views/categories/index.html.haml [hunk 21]
```diff
@@ -0,0 +1,10 @@
+%h1 Categories
```
Comment (vikabelous) [inline app/views/categories/index.html.haml]: Indentations

File: spec/controllers/categories_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+    it 'routes /categories/1 to categories#update' do
+      expect(put: '/categories/1').to route_to(controller: 'categories', action: 'update', id: '1')
+    end
+
+    it 'routes /categories/1 to categories#destroy' do
+      expect(delete: '/categories/1').to route_to(controller: 'categories',
+                                                  action: 'destroy', id: '1')
+    end
+  end
+  describe ' GET index ' do
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: Add empty lines between describe and context block, it's hard to read

File: spec/controllers/categories_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+  end
+  describe 'DELETE #destroy' do
+    before do
+      user = create(:user)
+      sign_in(user)
+    end
+    context 'DELETE #destroy' do
+      let!(:category) { create :category }
+      it 'should delete category' do
+        expect { delete :destroy, params: { id: category.id } }.to change(Category, :count).by(-1)
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: One expectation per test case

File: spec/helpers/categories_helper_spec.rb [hunk 24]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
```
Comment (vikabelous) [inline spec/helpers/categories_helper_spec.rb]: Do not use generators
Do not commit empty spec files, if there is nothing to test then delete this file

File: spec/rails_helper.rb [hunk 25]
```diff
@@ -15,6 +15,10 @@
 # Prevent database truncation if the environment is production
 abort('The Rails environment is running in production mode!') if Rails.env.production?
 require 'rspec/rails'
+require 'simplecov'
+require 'devise'
+SimpleCov.start
```
Comment (vikabelous) [inline spec/rails_helper.rb]: There is already SimpleCov required and started on lines 3-7

File: db/migrate/20190413204542_create_categories.rb (hunk 8 above)
Comment (nataliakozoriz) [reply db/migrate/20190413204542_create_categories.rb]: Changed model column name to description

File: config/database.yml.example (hunk 9 above)
Comment (nataliakozoriz) [reply config/database.yml.example]: Done

**Comment by vikabelous** [comment]
**Code Coverage:**
:warning: **88.1%** covered at **1.4** hits/line

File: app/helpers/categories_helper.rb (hunk 4 above)
Comment (nataliakozoriz) [reply app/helpers/categories_helper.rb]: Transferred to controller.

File: app/controllers/categories_controller.rb (hunk 5 above)
Comment (nataliakozoriz) [reply app/controllers/categories_controller.rb]: Used filter to assign variable.

File: spec/factories/users.rb (hunk 16 above)
Comment (nataliakozoriz) [reply spec/factories/users.rb]: Fixed.

File: spec/factories/users.rb (hunk 17 above)
Comment (nataliakozoriz) [reply spec/factories/users.rb]: Fixed.

File: app/assets/stylesheets/application.scss (hunk 18 above)
Comment (nataliakozoriz) [reply app/assets/stylesheets/application.scss]: Done.

File: app/models/category.rb (hunk 19 above)
Comment (nataliakozoriz) [reply app/models/category.rb]: Added validation to model to prevent this.

File: spec/controllers/categories_controller_spec.rb (hunk 22 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Done.

File: spec/helpers/categories_helper_spec.rb (hunk 24 above)
Comment (nataliakozoriz) [reply spec/helpers/categories_helper_spec.rb]: Deleted file.

File: spec/rails_helper.rb (hunk 25 above)
Comment (nataliakozoriz) [reply spec/rails_helper.rb]: Deleted those lines.

File: spec/controllers/categories_controller_spec.rb (hunk 23 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Deleted extra expectations.

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **92.86%** covered at **1.47** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **92.86%** covered at **1.47** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **92.86%** covered at **1.47** hits/line

File: app/views/categories/edit.html.haml (hunk 20 above)
Comment (nataliakozoriz) [reply app/views/categories/edit.html.haml]: Linter is not working. Fixed by hands.

File: app/views/categories/index.html.haml (hunk 21 above)
Comment (nataliakozoriz) [reply app/views/categories/index.html.haml]: Added indentations.

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **92.86%** covered at **1.47** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:warning: **84.62%** covered at **1.21** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:warning: **82.98%** covered at **1.12** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:warning: **82.98%** covered at **1.12** hits/line

File: app/models/category.rb [hunk 26]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class Category < ApplicationRecord
+  has_many :news
+
+  validates :name, presence: true, length: { minimum: 2 }, uniqueness: { scope: :name }
```
Comment (vikabelous) [inline app/models/category.rb]: Why`{ scope: :name }`?

File: app/controllers/categories_controller.rb [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+    if @category.save
+      redirect_to categories_path, flash: { success: 'Category was successfully created.' }
+    else
+      render 'new'
+    end
+  end
+
+  def destroy
+    if @category.destroy
+      redirect_to categories_path, flash: { success: 'Delete success.' }
```
Comment (vikabelous) [inline app/controllers/categories_controller.rb]: 'Category was successfully deleted'

File: app/controllers/categories_controller.rb [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+    else
+      render 'destroy'
+    end
+  end
+
+  def edit; end
+
+  def update
+    if @category.update(category_params)
+      redirect_to category_path(@category), flash: { success: 'Updated success.' }
```
Comment (vikabelous) [inline app/controllers/categories_controller.rb]: 'Category was successfully updated'

File: db/migrate/20190413204542_create_categories.rb [hunk 29]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateCategories < ActiveRecord::Migration[5.2]
+  def change
+    create_table :categories do |t|
+      t.string :name
```
Comment (vikabelous) [inline db/migrate/20190413204542_create_categories.rb]: You should add a unique index for `name` column
You have an app level uniqueness validation, but it's still possible to create 2 categories with the same name on the database level
https://thoughtbot.com/blog/the-perils-of-uniqueness-validations

File: db/schema.rb [hunk 30]
```diff
@@ -15,23 +15,41 @@
...
   enable_extension "plpgsql"
 
+  create_table "categories", force: :cascade do |t|
+    t.string "name"
+    t.text "description"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+  end
+
+  create_table "news", force: :cascade do |t|
```
Comment (vikabelous) [inline db/schema.rb]: These changes are not related to your PR, it shouldn't be there

File: db/schema.rb [hunk 31]
```diff
@@ -15,23 +15,41 @@
...
-    t.string "first_name", null: false
-    t.string "last_name", null: false
-    t.text "about_me"
     t.string "confirmation_token"
     t.datetime "confirmed_at"
     t.datetime "confirmation_sent_at"
     t.string "unconfirmed_email"
     t.string "reset_password_token"
     t.datetime "reset_password_sent_at"
+    t.string "first_name", null: false
```
Comment (vikabelous) [inline db/schema.rb]: Same as above

File: spec/controllers/categories_controller_spec.rb [hunk 32]
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'factory_bot'
+
+RSpec.describe CategoriesController, type: :controller do
+  before do
+    user = create(:user)
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: If you need user object in tests then use `let`
If it's just for authentication then this variable is redundant

File: spec/controllers/categories_controller_spec.rb [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+    sign_in(user)
+  end
+
+  describe '#index' do
+    it 'routes /categories to categories#index' do
+      expect(get: '/categories').to route_to(controller: 'categories', action: 'index')
+    end
+    it 'populates an array of categories' do
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: Add empty lines between `it` blocks

File: spec/controllers/categories_controller_spec.rb [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+      category = create(:category)
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: Use `let`, you already have one for destroy test

File: spec/controllers/categories_controller_spec.rb [hunk 35]
```diff
@@ -0,0 +1,68 @@
...
+    expect(assigns(:category)).to be_a_new(Category)
+  end
+
+  describe '#create'
+  it 'routes /categories to categories#create' do
+    expect(post: '/categories').to route_to(controller: 'categories', action: 'create')
+  end
+  it 'creates a category' do
+    params = attributes_for(:category)
+    expect { post(:create, params: { category: params }) }.to change(Category, :count).by(1)
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: It's not a good approach to test creation/deletion by count. Better use `exists?` method here and for `destroy`
https://apidock.com/rails/ActiveRecord/FinderMethods/exists%3F

File: spec/factories/categories.rb [hunk 36]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :category do
+    name { 'MyCategory' }
```
Comment (vikabelous) [inline spec/factories/categories.rb]: You would need a sequence for name attribute

File: app/controllers/categories_controller.rb (hunk 27 above)
Comment (nataliakozoriz) [reply app/controllers/categories_controller.rb]: Done.

File: app/controllers/categories_controller.rb (hunk 28 above)
Comment (nataliakozoriz) [reply app/controllers/categories_controller.rb]: Done.

File: db/schema.rb (hunk 30 above)
Comment (nataliakozoriz) [reply db/schema.rb]: Deleted unnessesary table.

File: spec/controllers/categories_controller_spec.rb (hunk 32 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Created let variables.

File: spec/factories/categories.rb (hunk 36 above)
Comment (nataliakozoriz) [reply spec/factories/categories.rb]: Done.

File: app/models/category.rb (hunk 26 above)
Comment (nataliakozoriz) [reply app/models/category.rb]: deleted unnecessary scope

File: db/migrate/20190413204542_create_categories.rb (hunk 29 above)
Comment (nataliakozoriz) [reply db/migrate/20190413204542_create_categories.rb]: Fixed migration and added unique to db

File: spec/controllers/categories_controller_spec.rb (hunk 33 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Done.

File: spec/controllers/categories_controller_spec.rb (hunk 34 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Deleted.

File: spec/controllers/categories_controller_spec.rb (hunk 35 above)
Comment (nataliakozoriz) [reply spec/controllers/categories_controller_spec.rb]: Changed to exists

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.48%** covered at **1.98** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.48%** covered at **1.98** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.88%** covered at **2.09** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.88%** covered at **2.09** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.88%** covered at **2.09** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.88%** covered at **2.09** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.23%** covered at **2.06** hits/line

File: app/assets/stylesheets/categories.scss [hunk 37]
```diff
@@ -0,0 +1,46 @@
...
+ *= require_self
+ */
+
+$white-color: #ffffff;
+
+.category {
+  font-size: 24px;
+}
+
+input[type="submit"] {
```
[the remark this answers is not in the document]
Comment (nataliakozoriz) [reply app/assets/stylesheets/categories.scss]: I need to use it here

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.23%** covered at **2.06** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.23%** covered at **2.06** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.23%** covered at **2.06** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.23%** covered at **2.06** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.23%** covered at **2.06** hits/line

File: app/assets/stylesheets/categories.scss [hunk 38]
```diff
@@ -0,0 +1,6 @@
+// = require_tree .
+// = require_self
```
Comment (okohuch) [inline app/assets/stylesheets/categories.scss]: these two should not be here

File: app/controllers/categories_controller.rb [hunk 39]
```diff
@@ -0,0 +1,56 @@
...
+      redirect_to @category, flash: { success: 'Category was successfully updated.' }
+    else
+      render 'edit', alert: 'The edit was unsuccessful'
+    end
+  end
+
+  private
+
+  def set_category
+    @category = Category.find_by(id: params[:id])
```
Comment (okohuch) [inline app/controllers/categories_controller.rb]: change `find_by` to `find`

File: app/controllers/categories_controller.rb [hunk 40]
```diff
@@ -0,0 +1,56 @@
...
+
+    if @category.save
+      redirect_to categories_path, flash: { success: 'Category was successfully created.' }
+    else
+      render 'new', alert: 'The category was not saved'
+    end
+  end
+
+  def destroy
+    if @category.present?
```
Comment (okohuch) [inline app/controllers/categories_controller.rb]: please work with @kmandziy

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **92.16%** covered at **2.1** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **92.16%** covered at **2.1** hits/line

File: app/assets/stylesheets/categories.scss [hunk 41]
```diff
@@ -0,0 +1,3 @@
+.category {
```
Comment (vikabelous) [inline app/assets/stylesheets/categories.scss]: You have the same style in `application.scss`

File: db/schema.rb (hunk 30 above)
Comment (vikabelous) [reply db/schema.rb]: Your schema still has `news` table but it was not merged yet

File: db/schema.rb [hunk 42]
```diff
@@ -15,24 +15,45 @@
...
+    t.datetime "updated_at", null: false
+    t.index ["user_id"], name: "index_news_on_user_id"
+  end
+
   create_table "users", force: :cascade do |t|
     t.string "email", default: "", null: false
     t.string "encrypted_password", default: "", null: false
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.string "first_name", null: false
```
Comment (vikabelous) [inline db/schema.rb]: And these changes also not related to your PR

File: spec/models/category_spec.rb [hunk 43]
```diff
@@ -0,0 +1,24 @@
...
+require 'factory_bot'
+
+RSpec.describe Category, type: :model do
+  context 'when all parameters are correct' do
+    it 'is valid' do
+      expect(build(:category)).to be_valid
+    end
+  end
+
+  context 'when name is blank' do
```
Comment (vikabelous) [inline spec/models/category_spec.rb]: Add test for validation of uniquness

File: spec/models/category_spec.rb [hunk 44]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'factory_bot'
```
Comment (vikabelous) [inline spec/models/category_spec.rb]: Do you really need to require this?

File: spec/controllers/categories_controller_spec.rb [hunk 45]
```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'factory_bot'
+
+RSpec.describe CategoriesController, type: :controller do
+  let(:params) do
+    {
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: Use `attributes_for`

File: spec/controllers/categories_controller_spec.rb [hunk 46]
```diff
@@ -0,0 +1,130 @@
...
+      name: 'new name',
+      description: 'new description'
+    }
+  end
+  let(:user) { create(:user) }
+  let!(:category) { create :category }
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: Please be consistent at least in scope of your own changes
line 13 `create(:user)`
line 14 `create :category`

File: spec/controllers/categories_controller_spec.rb [hunk 47]
```diff
@@ -0,0 +1,130 @@
...
+  let(:invalid_data) { attributes_for(:category, name: '') }
+  before do
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: Add empty line before `before` block

File: spec/controllers/categories_controller_spec.rb [hunk 48]
```diff
@@ -0,0 +1,130 @@
...
+    end
+  end
+
+  describe '#create' do
+    it 'routes /categories to categories#create' do
+      expect(post: '/categories').to route_to(controller: 'categories', action: 'create')
+    end
+
+    it 'creates a category with valid data' do
+      expect(Category.exists?(category.id)).to be_truthy
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: You don't test controller here
Your `category` is created in `let` and not via the `create` request

File: spec/controllers/categories_controller_spec.rb [hunk 49]
```diff
@@ -0,0 +1,130 @@
...
+    end
+
+    it 'redirects to categories#show' do
+      post :create, params: { id: category.id, category: params }
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: Why do you pass the `id` to the create action?

File: spec/controllers/categories_controller_spec.rb [hunk 50]
```diff
@@ -0,0 +1,130 @@
...
+      expect(put: '/categories/1').to route_to(controller: 'categories', action: 'update', id: '1')
+    end
+
+    it 'updates category' do
+      put :update, params: { id: category.id, category: params }
+      category.reload
+      expect(category.name).to eq params[:name]
+    end
+
+    it 'renders :edit template if data is invalid' do
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: When you have `if` in your `it` description, it's a sign that you should use `context` block

File: spec/controllers/categories_controller_spec.rb [hunk 51]
```diff
@@ -0,0 +1,130 @@
...
+      expect(delete: '/categories/1').to route_to(controller: 'categories',
+                                                  action: 'destroy', id: '1')
+    end
+
+    it 'deletes achievements from database' do
+      delete :destroy, params: { id: category.id }
+      expect(Category.exists?(category.id)).to be_falsy
+    end
+
+    it 'does not delete news' do
```
Comment (vikabelous) [inline spec/controllers/categories_controller_spec.rb]: If you run this spec file with `--format documentation` it will print next output
```
#destroy
  routes /categories/1 to categories#destroy
  deletes achievements from database
  does not delete news
```

This looks confusing

File: app/controllers/categories_controller.rb [hunk 52]
```diff
@@ -0,0 +1,55 @@
...
+
+  def new
+    @category = Category.new
+  end
+
+  def create
+    @category = Category.new(category_params)
+
+    if @category.save
+      redirect_to categories_path, flash: { success: 'Category was successfully created.' }
```
Comment (vikabelous) [inline app/controllers/categories_controller.rb]: You use `success`, Khrystyna uses `notice`
https://github.com/dark-side/ai-collaboration-platform/pull/9/files#diff-34c11781522250e8269253c98cd1e623R24

Talk to @kmandziy and consider a single approach

File: app/views/categories/new.html.haml [hunk 53]
```diff
@@ -0,0 +1,16 @@
+.container
```
Comment (vikabelous) [inline app/views/categories/new.html.haml]: This template is almost identical to the `edit`
Add `form` partial for shared logic
https://github.com/dark-side/ai-collaboration-platform/pull/9/files#diff-44c13d674a5e814aca0c602a0c8e53cc

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.33%** covered at **2.57** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **95.6%** covered at **2.74** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **95.6%** covered at **2.74** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **95.6%** covered at **2.74** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **95.6%** covered at **2.74** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **95.6%** covered at **2.74** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **95.6%** covered at **2.74** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **95.6%** covered at **2.74** hits/line
